mitodl/mit-open
# Add filtering for topics, tags, level

# What are the relevant tickets?
Closes #281

# Description (What does it do?)
- Adds `/api/v1/learning_resources` filters for level, topic and resource_content_tags.
- Modifies `LearningResourceRun.level` to be an ArrayField


# Screenshots (if appropriate):

<img width="849" alt="Screenshot 2023-12-13 at 11 50 48 AM" src="https://github.com/mitodl/mit-open/assets/187676/773c10e8-e934-4f5d-b212-d9cf94b5e393">
<img width="850" alt="Screenshot 2023-12-13 at 11 50 41 AM" src="https://github.com/mitodl/mit-open/assets/187676/d9fb424c-4d52-41e2-b4ee-7dc828a3721d">
<img width="587" alt="Screenshot 2023-12-13 at 11 50 32 AM" src="https://github.com/mitodl/mit-open/assets/187676/4167acbd-e285-4a89-9359-8c0c7c8ef23d">


# How can this be tested?

The migration should complete successfully with existing learning resource data present.

The following urls should work and return relevant results (assuming you've imported some matching courses from ocw, mostly).

http://localhost:8063/api/v1/learning_resources/?resource_content_tags=exams
http://localhost:8063/api/v1/learning_resources/?topic=social+science
http://localhost:8063/api/v1/learning_resources/?level=undergraduate

Run the following, they should complete without errors (you'll need the approproate .env values set first):

```
./manage.py backpopulate_mit_edx_data
./manage.py backpopulate_ocw_data --course-name 8-591j-systems-biology-fall-2014
```



# Additional Context
I avoided specifying the choices for resource_content_tags and topics because doing so required querying the database on application load, which basically breaks pytest, [requiring hacks to get around](https://github.com/pytest-dev/pytest-django/issues/396#issuecomment-393444718). Instead, I also tried adding them to the `LearningResourceViewSet` via `OpenApiParameter` specs instead ([in this commit](https://github.com/mitodl/mit-open/pull/282/commits/f08eb9d59185f7279cac6cb915b7e16316dec141#diff-12658d780aa8728f3af111a6a3fb4e1c1fe224e462c74f8bc62555b42f258ef9R86-R158)), but that caused the ci tests to fail because the database was empty and the api specs didn't match what was generated locally with data. So I ultimately went back to my original approach of just noting that the topics and resource_content_tags api endpoints can be used to get the list of available values, instead of listing all of them.  (The resource_content_tags endpoint is being added in a separate PR).

As for levels, I decided to use an Enum.
